File: 2025-09-September-GFG-POTD/September 21 - Max rectangle/Solution.py
class Solution:
    def maxArea(self, mat):
        if not mat or not mat[0]:
            return 0

        n, m = len(mat), len(mat[0])
        heights = [0] * m
        max_area = 0

        for i in range(n):
            # update histogram heights
            for j in range(m):
                if mat[i][j] == 0:
                    heights[j] = 0
                else:
                    heights[j] += 1

            # largest rectangle in histogram
            stack = []
            for j in range(m + 1):
                h = 0 if j == m else heights[j]
                while stack and h < heights[stack[-1]]:
                    height = heights[stack.pop()]
                    width = j if not stack else j - stack[-1] - 1
                    max_area = max(max_area, height * width)
                stack.append(j)

        return max_area


# Example usage:
# mat = [[0,1,1,0],
#        [1,1,1,1],
#        [1,1,1,1],
#        [1,1,0,0]]
# sol = Solution()
# print(sol.maxArea(mat))  # Output: 8

File: 2025-09-September-GFG-POTD/September 21 - Max rectangle/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxArea(vector<vector<int>>& mat) {
        int n = mat.size(), m = mat[0].size();
        vector<int> heights(m, 0);
        int maxArea = 0;

        for (int i = 0; i < n; i++) {
            // update histogram heights
            for (int j = 0; j < m; j++) {
                if (mat[i][j] == 0) heights[j] = 0;
                else heights[j]++;
            }

            // largest rectangle in histogram
            stack<int> st;
            for (int j = 0; j <= m; j++) {
                int h = (j == m ? 0 : heights[j]);
                while (!st.empty() && h < heights[st.top()]) {
                    int height = heights[st.top()];
                    st.pop();
                    int width = st.empty() ? j : j - st.top() - 1;
                    maxArea = max(maxArea, height * width);
                }
                st.push(j);
            }
        }
        return maxArea;
    }
};

// Example usage:
// int main() {
//     vector<vector<int>> mat = {{0,1,1,0},{1,1,1,1},{1,1,1,1},{1,1,0,0}};
//     Solution sol;
//     cout << sol.maxArea(mat) << endl; // Output: 8
// }
